dali/imgcodec/decoders/nvjpeg_lossless/nvjpeg_lossless.h

```diff
@@ -42,20 +42,39 @@ class DLL_PUBLIC NvJpegLosslessDecoderInstance : public BatchedApiDecoderImpl {
                                      cspan<ROI> rois = {}) override;
 
  private:
+  // Parses encoded streams and populates SampleMeta, and batch_sz_
+  void Parse(DecodeResultsPromise &promise, DecodeContext ctx, cspan<ImageSource *> in,
+             DecodeParams opts, cspan<ROI> rois);
+
+  // Invokes nvJPEG decoding (sample_meta_ and batch_sz_ to be populated)
+  void RunDecode(DecodeContext ctx, span<SampleView<GPUBackend>> out,
+                 cspan<ImageSource *> in, DecodeParams opts,
+                 cspan<ROI> rois = {});
+
   void Postprocess(DecodeResultsPromise &promise, DecodeContext ctx,
                    span<SampleView<GPUBackend>> out, DecodeParams opts, cspan<ROI> rois);
 
   nvjpegHandle_t nvjpeg_handle_;
+
+  struct PerThreadResources {
+    explicit PerThreadResources(nvjpegHandle_t handle);
+    PerThreadResources(PerThreadResources&& other);
+    ~PerThreadResources();
+    nvjpegJpegStream_t jpeg_stream;
+  };
+  std::vector<PerThreadResources> per_thread_resources_;
   nvjpegJpegStream_t jpeg_stream_;
   CUDAEvent event_;
   nvjpegJpegState_t state_;
 
+  int batch_sz_ = 0;  // number of samples to be decoded by nvJPEG
```

Comment:
This isn't necessary - you could take it from `sample_meta_.size()` - resizing a vector _down_ is cheap.